app/domain/authentication/authn_azure/application_identity.rb

```diff
@@ -0,0 +1,41 @@
+module Authentication
+  module AuthnAzure
+
+    Log = LogMessages::Authentication::AuthnAzure
+
+    class ApplicationIdentity
+
+      def initialize(role_annotations:, service_id:)
+        @role_annotations = role_annotations
+        @service_id       = service_id
+      end
+
+      def constraints
+        @constraints ||= {
+          subscription_id:          constraint_value("subscription-id"),
+          resource_group:           constraint_value("resource-group"),
+          user_assigned_identity:   constraint_value("user-assigned-identity"),
+          system_assigned_identity: constraint_value("system-assigned-identity")
+        }.compact
+      end
+
+      private
+
+      # check the `service-id` specific constraint first to be more granular
+      def constraint_value constraint_name
+        annotation_value("authn-azure/#{@service_id}/#{constraint_name}") ||
+          annotation_value("authn-azure/#{constraint_name}")
```

Comment:
Without looking at this more deeply, something strikes me as off about having two ways to store the annotation.  It might be ok if this logic only exists in this one place.  Otherwise, it should likely be moved to an object of its own.

Reply:
Thanks, let's tackle this in the application identity refactor